### PR TITLE
Week 5

### PR DESCRIPTION
**Pull request by mz-evgeny**
[link to  deploy:](https://gallant-mayer-81f65d.netlify.app/)
Done:
- add redux, redux-thunk (for Member page);
- add settings with Themes mode (3 themes);
- updated to react 18.

user data for login:
Admin:
email: [admin@gmail.com](mailto:admin@gmail.com)
password: qwer1234
Mentor
email: [tolya@gmail.com](mailto:tolya@gmail.com)
password: qwer1234
Member
email: [yana@gmail.com](mailto:yana@gmail.com)
password: qwer1234

### CODE REVIEW

File: src/Hooks/useSettings.jsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import PropTypes from 'prop-types';
+
+export const SettingsContext = createContext(null);
+
+export class SettingsProvider extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.state = {
+      theme: {
+        isShowSettingsPage: false,
```
Comment (maximsan) [inline src/Hooks/useSettings.jsx]: `isShowSettings` is more than enough for name

File: src/Hooks/useSettings.jsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        toggleSettingsPageHandler: this.toggleSettingsPage,
```
Comment (maximsan) [inline src/Hooks/useSettings.jsx]: `toggleSettings` is more than enough for name

File: src/components/Common/SettingsPage/SetingsPage.jsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  const { body } = document;
+  const setLightTheme = () => {
+    body.setAttribute('data-theme', 'light');
+  };
+  const setDarkTheme = () => {
+    body.setAttribute('data-theme', 'dark');
+  };
+  const setDarkDimmedTheme = () => {
+    body.setAttribute('data-theme', 'dark-dimmed');
+  };
```
Comment (maximsan) [inline src/components/Common/SettingsPage/SetingsPage.jsx]: These methods should be in ThemeContext and create one `toggleTheme(theme)` method.

File: src/components/Pages/Members/Members.jsx [hunk 4]
```diff
@@ -194,3 +197,32 @@ export class Members extends React.PureComponent {
...
+    users: state.users.users,
+  };
+};
+
+const mapDispatchToProps = (dispatch) => {
+  return {
+    getUsers: () => dispatch(getUsersThunk()),
+    removeUser: (userId) => dispatch(removeUserThunk(userId)),
+    editUser: (userId, userData) => dispatch(editUserThunk(userId, userData)),
+    createUser: (userData) => dispatch(createUserThunk(userData)),
```
Comment (maximsan) [inline src/components/Pages/Members/Members.jsx]: use `bindActionCreators` to make code more concise

File: src/store/actionCreators/usersActionCreators.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import { getAllUsers, removeUserData, editUser, createUser } from '../../services/users-services ';
+import { EDIT_USER, GET_USERS, REMOVE_USER } from '../actions/actions';
+
+const getUsers = (users) => ({ type: GET_USERS, payload: { users } });
+const removeUser = (userId) => ({ type: REMOVE_USER, payload: { userId } });
+const editUserData = (userId, userData) => ({ type: EDIT_USER, payload: { userId, userData } });
+
+export function getUsersThunk() {
+  return async (dispatch) => {
+    const users = await getAllUsers();
```
Comment (maximsan) [inline src/store/actionCreators/usersActionCreators.js:10]: where do you handler possible errors ? I don't see it.

**Review by maximsan — CHANGES_REQUESTED** [review]
Hello! I left some comments. Please fix and let me know if you have questions. Thanks !

File: src/components/Pages/Members/Members.jsx (hunk 4 above)
Comment (mz-evgeny) [reply src/components/Pages/Members/Members.jsx]:
```suggestion
    getUsers: bindActionCreators(getUsersThunk, dispatch),
    removeUser: bindActionCreators(removeUserThunk, dispatch),
    editUser: bindActionCreators(editUserThunk, dispatch),
    createUser: bindActionCreators(createUserThunk, dispatch),
```

fixed

File: src/Hooks/useSettings.jsx (hunk 1 above)
Comment (LeoKuD) [reply src/Hooks/useSettings.jsx:11]: fixed

Comment (mz-evgeny) [reply src/Hooks/useSettings.jsx]:
```suggestion
        isShowSettings: false,
```

fixed

File: src/Hooks/useSettings.jsx (hunk 2 above)
Comment (mz-evgeny) [reply src/Hooks/useSettings.jsx]:
```suggestion
        toggleSettings: this.toggleSettings,
```
fixed

File: src/components/Common/SettingsPage/SetingsPage.jsx (hunk 3 above)
Comment (mz-evgeny) [reply src/components/Common/SettingsPage/SetingsPage.jsx]:
```
  toggleTheme(theme) {
    this.setState((prevState) => ({ theme: { ...prevState.theme, theme } }));
  }
```

I named this Context so SettingsContext because I can add still more settengs. in general, I understand you.
Сan i leave it like this? Thanks.

File: src/store/actionCreators/usersActionCreators.js (hunk 5 above)
Comment (mz-evgeny) [reply src/store/actionCreators/usersActionCreators.js:10]: Error handling in getAllUsers(). Should I process them here too?

File: src/components/Pages/Members/Members.jsx (hunk 4 above)
Comment (maximsan) [reply src/components/Pages/Members/Members.jsx]: you can write like `bindActionCreators({removeUser: removeUserThunk, editUser: editUserThunk}, dispatch)` as far as I remember

File: src/store/actionCreators/usersActionCreators.js (hunk 5 above)
Comment (maximsan) [reply src/store/actionCreators/usersActionCreators.js:10]: if it's handled there than you can omit it here

File: src/Hooks/useSettings.jsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  constructor(props) {
+    super(props);
+    this.state = {
+      theme: {
+        isShowSettings: false,
+        toggleSettings: this.toggleSettings,
+        setLightTheme: this.toggleTheme.bind(this, 'light'),
+        setDarkTheme: this.toggleTheme.bind(this, 'dark'),
+        setDarkDimmedTheme: this.toggleTheme.bind(this, 'dark-dimmed'),
+      },
```
Comment (maximsan) [inline src/Hooks/useSettings.jsx]: you violate several principles here, because these all fields are not related to theme object, I'd recommend to remove theme object at all and just pass a state with all fields directly

**Review by maximsan — APPROVED** [review]
ok, I left some comments, please fix. It's approved.

File: src/components/Pages/Members/Members.jsx (hunk 4 above)
Comment (mz-evgeny) [reply src/components/Pages/Members/Members.jsx]:
```
const mapDispatchToProps = (dispatch) => {
  return bindActionCreators(
    { getUsers: getUsersThunk, removeUser: removeUserThunk, editUser: editUserThunk, createUser: createUserThunk },
    dispatch,
  );
};
```
fix, thanks

File: src/Hooks/useSettings.jsx (hunk 6 above)
Comment (mz-evgeny) [reply src/Hooks/useSettings.jsx]:
```
    this.state = {
      isShowSettings: false,
      toggleSettings: this.toggleSettings,
      setLightTheme: this.toggleTheme.bind(this, 'light'),
      setDarkTheme: this.toggleTheme.bind(this, 'dark'),
      setDarkDimmedTheme: this.toggleTheme.bind(this, 'dark-dimmed'),
    };
```
Fixed.
I don't remember why I couldn't do it before. Thanks.

**Review by oleg3555 — APPROVED** [review]
Have a good week, approved)
